README.md
A simple JSON Parser / builder
------------------------------

SimpleJSON mainly has been written as a simple JSON parser. It can build a JSON string
from the node-tree, or generate a node tree from any valid JSON string.

Written by Bunny83
2012-06-09

Add the following to your package.json "dependencies" section to import SimpleJSON:
    "com.github.bunny83.simplejson": "https://github.com/xeptor6569/F.SimpleJSON.git",

SimpleJSONBinary is an extension of the SimpleJSON framework to provide methods to
serialize a JSON object tree into a compact binary format. Optionally the
binary stream can be compressed with the SharpZipLib when using the define
"USE_SharpZipLib"

Those methods where originally part of the framework but since it's rarely
used I've extracted this part into this seperate module file.

You can use the define "SimpleJSON_ExcludeBinary" to selectively disable
this extension without the need to remove the file from the project.

If you want to use compression when saving to file / stream / B64 you have to include
SharpZipLib ( http://www.icsharpcode.net/opensource/sharpziplib/ ) in your project and
define "USE_SharpZipLib" at the top of the file

SimpleJSONUnity is a Unity extension for the SimpleJSON framework. It does
only work together with the SimpleJSON.cs
It provides several helpers and conversion operators to serialize/deserialize
common Unity types such as Vector2/3/4, Rect, RectOffset, Quaternion and
Matrix4x4 as JSONObject or JSONArray.
This extension will add 3 static settings to the JSONNode class:
( VectorContainerType, QuaternionContainerType, RectContainerType ) which
control what node type should be used for serializing the given type. So a
Vector3 as array would look like [12,32,24] and {"x":12, "y":32, "z":24} as
object.


CHANGELOG.md
/*
 * [2012-06-09 First Version]
 * - provides strongly typed node classes and lists / dictionaries
 * - provides easy access to class members / array items / data values
 * - the parser now properly identifies types. So generating JSON with this framework should work.
 * - only double quotes (") are used for quoting strings.
 * - provides "casting" properties to easily convert to / from those types:
 *   int / float / double / bool
 * - provides a common interface for each node so no explicit casting is required.
 * - the parser tries to avoid errors, but if malformed JSON is parsed the result is more or less undefined
 * - It can serialize/deserialize a node tree into/from an experimental compact binary format. It might
 *   be handy if you want to store things in a file and don't want it to be easily modifiable
 * 
 * [2012-12-17 Update]
 * - Added internal JSONLazyCreator class which simplifies the construction of a JSON tree
 *   Now you can simple reference any item that doesn't exist yet and it will return a JSONLazyCreator
 *   The class determines the required type by it's further use, creates the type and removes itself.
 * - Added binary serialization / deserialization.
 * - Added support for BZip2 zipped binary format. Requires the SharpZipLib ( http://www.icsharpcode.net/opensource/sharpziplib/ )
 *   The usage of the SharpZipLib library can be disabled by removing or commenting out the USE_SharpZipLib define at the top
 * - The serializer uses different types when it comes to store the values. Since my data values
 *   are all of type string, the serializer will "try" which format fits best. The order is: int, float, double, bool, string.
 *   It's not the most efficient way but for a moderate amount of data it should work on all platforms.
 * 
 * [2017-03-08 Update]
 * - Optimised parsing by using a StringBuilder for token. This prevents performance issues when large
 *   string data fields are contained in the json data.
 * - Finally refactored the badly named JSONClass into JSONObject.
 * - Replaced the old JSONData class by distict typed classes ( JSONString, JSONNumber, JSONBool, JSONNull ) this
 *   allows to propertly convert the node tree back to json without type information loss. The actual value
 *   parsing now happens at parsing time and not when you actually access one of the casting properties.
 * 
 * [2017-04-11 Update]
 * - Fixed parsing bug where empty string values have been ignored.
 * - Optimised "ToString" by using a StringBuilder internally. This should heavily improve performance for large files
 * - Changed the overload of "ToString(string aIndent)" to "ToString(int aIndent)"
 * 
 * [2017-11-29 Update]
 * - Removed the IEnumerator implementations on JSONArray & JSONObject and replaced it with a common
 *   struct Enumerator in JSONNode that should avoid garbage generation. The enumerator always works
 *   on KeyValuePair<string, JSONNode>, even for JSONArray.
 * - Added two wrapper Enumerators that allows for easy key or value enumeration. A JSONNode now has
 *   a "Keys" and a "Values" enumerable property. Those are also struct enumerators / enumerables
 * - A KeyValuePair<string, JSONNode> can now be implicitly converted into a JSONNode. This allows
 *   a foreach loop over a JSONNode to directly access the values only. Since KeyValuePair as well as
 *   all the Enumerators are structs, no garbage is allocated.
 * - To add Linq support another "LinqEnumerator" is available through the "Linq" property. This
 *   enumerator does implement the generic IEnumerable interface so most Linq extensions can be used
 *   on this enumerable object. This one does allocate memory as it's a wrapper class.
 * - The Escape method now escapes all control characters (# < 32) in strings as uncode characters
 *   (\uXXXX) and if the static bool JSONNode.forceASCII is set to true it will also escape all
 *   characters # > 127. This might be useful if you require an ASCII output. Though keep in mind
 *   when your strings contain many non-ascii characters the strings become much longer (x6) and are
 *   no longer human readable.
 * - The node types JSONObject and JSONArray now have an "Inline" boolean switch which will default to
 *   false. It can be used to serialize this element inline even you serialize with an indented format
 *   This is useful for arrays containing numbers so it doesn't place every number on a new line
 * - Extracted the binary serialization code into a seperate extension file. All classes are now declared
 *   as "partial" so an extension file can even add a new virtual or abstract method / interface to
 *   JSONNode and override it in the concrete type classes. It's of course a hacky approach which is
 *   generally not recommended, but i wanted to keep everything tightly packed.
 * - Added a static CreateOrGet method to the JSONNull class. Since this class is immutable it could
 *   be reused without major problems. If you have a lot null fields in your data it will help reduce
 *   the memory / garbage overhead. I also added a static setting (reuseSameInstance) to JSONNull
 *   (default is true) which will change the behaviour of "CreateOrGet". If you set this to false
 *   CreateOrGet will not reuse the cached instance but instead create a new JSONNull instance each time.
 *   I made the JSONNull constructor private so if you need to create an instance manually use
 *   JSONNull.CreateOrGet()
 * 
 * [2018-01-09 Update]
 * - Changed all double.TryParse and double.ToString uses to use the invariant culture to avoid problems
 *   on systems with a culture that uses a comma as decimal point.
 * 
 * [2018-01-26 Update]
 * - Added AsLong. Note that a JSONNumber is stored as double and can't represent all long values. However
 *   storing it as string would work.
 * - Added static setting "JSONNode.longAsString" which controls the default type that is used by the

Tests/Editor/SimpleJSONUnityTests.cs
using NUnit.Framework;
using XGE.SimpleJSON;
using UnityEngine;

namespace Tests
{
    public class SimpleJSONUnityTests
    {
        [Test]
        public void Vector2Test()
        {
            var vec2 = Random.insideUnitCircle;

            var jsonObject = new JSONObject().WriteVector2(vec2);
            var jsonArray = new JSONArray().WriteVector2(vec2);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{vec2.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadVector2();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadVector2();

            Assert.AreEqual(vec2, deserializedObject);
            Assert.AreEqual(vec2, deserializedArray);
        }

        [Test]
        public void Vector3Test()
        {
            var vec3 = Random.insideUnitSphere;

            var jsonObject = new JSONObject().WriteVector3(vec3);
            var jsonArray = new JSONArray().WriteVector3(vec3);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{vec3.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadVector3();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadVector3();

            Assert.AreEqual(vec3, deserializedObject);
            Assert.AreEqual(vec3, deserializedArray);
        }

        [Test]
        public void Vector4Test()
        {
            Vector4 vec4 = Random.insideUnitSphere;
            vec4.w = Random.value;

            var jsonObject = new JSONObject().WriteVector4(vec4);
            var jsonArray = new JSONArray().WriteVector4(vec4);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{vec4.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadVector4();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadVector4();

            Assert.AreEqual(vec4, deserializedObject);
            Assert.AreEqual(vec4, deserializedArray);
        }

        [Test]
        public void QuaternionTest()
        {
            Quaternion quat = Random.rotation;

            var jsonObject = new JSONObject().WriteQuaternion(quat);
            var jsonArray = new JSONArray().WriteQuaternion(quat);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{quat.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadQuaternion();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadQuaternion();

            Assert.AreEqual(quat, deserializedObject);
            Assert.AreEqual(quat, deserializedArray);
        }

        [Test]
        public void PoseTest()
        {
            Vector3 position = Random.insideUnitSphere * 256.0f;
            Quaternion rotation = Random.rotation;

            Pose pose = new Pose(position, rotation);

            var jsonObject = new JSONObject().WritePose(pose);
            var jsonArray = new JSONArray().WritePose(pose);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{pose.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadPose();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadPose();

            Assert.AreEqual(pose, deserializedObject);
            Assert.AreEqual(pose, deserializedArray);
        }

        [Test]
        public void RectTest()
        {
            Rect rect = new Rect(Random.insideUnitCircle * 10.0f, Random.insideUnitCircle * 10.0f);

            var jsonObject = new JSONObject().WriteRect(rect);
            var jsonArray = new JSONArray().WriteRect(rect);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{rect.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadRect();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadRect();

            Assert.AreEqual(rect, deserializedObject);
            Assert.AreEqual(rect, deserializedArray);
        }

        [Test]
        public void RectOffsetTest()
        {
            RectOffset rectOffset = new RectOffset(Random.Range(-100, 100), Random.Range(-100, 100), Random.Range(-100, 100), Random.Range(-100, 100));

            var jsonObject = new JSONObject().WriteRectOffset(rectOffset);
            var jsonArray = new JSONArray().WriteRectOffset(rectOffset);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{rectOffset.GetType().Name} object: {jsonObjectString} array: {jsonArrayString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadRectOffset();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadRectOffset();

            Assert.AreEqual(rectOffset.left, deserializedObject.left);
            Assert.AreEqual(rectOffset.right, deserializedObject.right);
            Assert.AreEqual(rectOffset.top, deserializedObject.top);
            Assert.AreEqual(rectOffset.bottom, deserializedObject.bottom);

            Assert.AreEqual(rectOffset.left, deserializedArray.left);
            Assert.AreEqual(rectOffset.right, deserializedArray.right);
            Assert.AreEqual(rectOffset.top, deserializedArray.top);
            Assert.AreEqual(rectOffset.bottom, deserializedArray.bottom);
        }

        [Test]
        public void MatrixTest()
        {
            Matrix4x4 matrix = new Matrix4x4();
            for (int i = 0; i < 16; i++)
            {
                matrix[i] = Random.Range(-1.0f, 1.0f);
            }

            var jsonArray = new JSONArray().WriteMatrix(matrix);

            var jsonArrayString = jsonArray.ToString();

            Debug.Log($"{matrix.GetType().Name} array: {jsonArrayString}");

            var deserializedArray = JSON.Parse(jsonArrayString).ReadMatrix();

            Assert.AreEqual(matrix, deserializedArray);
        }

        [Test]
        public void ColorTest()
        {
            Color color = new Color(Random.value, Random.value, Random.value, Random.value);

            var jsonObject = new JSONObject().WriteColor(color);
            var jsonArray = new JSONArray().WriteColor(color);
            var jsonString = new JSONString().WriteColor(color);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();
            var jsonStringString = jsonString.ToString();

            Debug.Log($"{color.GetType().Name} object: {jsonObjectString} array: {jsonArrayString} string: {jsonStringString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadColor();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadColor();

            Assert.AreEqual(color, deserializedObject);
            Assert.AreEqual(color, deserializedArray);
            Assert.AreEqual(deserializedObject, deserializedArray);
        }

        [Test]
        public void Color32Test()
        {
            Color32 color32 = new Color(Random.value, Random.value, Random.value, Random.value);

            var jsonObject = new JSONObject().WriteColor32(color32);
            var jsonArray = new JSONArray().WriteColor32(color32);
            var jsonString = new JSONString().WriteColor32(color32);

            var jsonObjectString = jsonObject.ToString();
            var jsonArrayString = jsonArray.ToString();
            var jsonStringString = jsonString.ToString();

            Debug.Log($"{color32.GetType().Name} object: {jsonObjectString} array: {jsonArrayString} string: {jsonStringString}");

            var deserializedObject = JSON.Parse(jsonObjectString).ReadColor32();
            var deserializedArray = JSON.Parse(jsonArrayString).ReadColor32();
            var deserializedString = JSON.Parse(jsonStringString).ReadColor32();

            Assert.AreEqual(color32, deserializedObject);
            Assert.AreEqual(color32, deserializedArray);
            Assert.AreEqual(color32, deserializedString);
            Assert.AreEqual(deserializedObject, deserializedArray);
            Assert.AreEqual(deserializedString, deserializedArray);
        }
    }
}


Runtime/SimpleJSONUnity.cs
#if UNITY_5_3_OR_NEWER
#region License and information
/* * * * *
 * 
 * Unity extension for the SimpleJSON framework. It does only work together with
 * the SimpleJSON.cs
 * It provides several helpers and conversion operators to serialize/deserialize
 * common Unity types such as Vector2/3/4, Rect, RectOffset, Quaternion and
 * Matrix4x4 as JSONObject or JSONArray.
 * This extension will add 3 static settings to the JSONNode class:
 * ( VectorContainerType, QuaternionContainerType, RectContainerType ) which
 * control what node type should be used for serializing the given type. So a
 * Vector3 as array would look like [12,32,24] and {"x":12, "y":32, "z":24} as
 * object.
 * 
 * 
 * The MIT License (MIT)
 * 
 * Copyright (c) 2012-2017 Markus Göbel (Bunny83)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * * * * */

#endregion License and information
using UnityEngine;

namespace XGE.SimpleJSON
{
    public enum JSONContainerType { Array, Object }
    public partial class JSONNode
    {
        public static JSONContainerType VectorContainerType = JSONContainerType.Array;
        public static JSONContainerType QuaternionContainerType = JSONContainerType.Array;
        public static JSONContainerType RectContainerType = JSONContainerType.Object;
        public static JSONContainerType ColorContainerType = JSONContainerType.Object;
        private static JSONNode GetContainer(JSONContainerType aType)
        {
            if (aType == JSONContainerType.Array)
            {
                return new JSONArray();
            }
            return new JSONObject();
        }

        #region implicit conversion operators
        public static implicit operator JSONNode(Vector2 aVec)
        {
            JSONNode n = GetContainer(VectorContainerType);
            n.WriteVector2(aVec);
            return n;
        }

        public static implicit operator JSONNode(Vector3 aVec)
        {
            JSONNode n = GetContainer(VectorContainerType);
            n.WriteVector3(aVec);
            return n;
        }

        public static implicit operator JSONNode(Vector4 aVec)
        {
            JSONNode n = GetContainer(VectorContainerType);
            n.WriteVector4(aVec);
            return n;
        }

        public static implicit operator JSONNode(Quaternion aRot)
        {
            JSONNode n = GetContainer(QuaternionContainerType);
            n.WriteQuaternion(aRot);
            return n;
        }

        public static implicit operator JSONNode(Rect aRect)
        {
            JSONNode n = GetContainer(RectContainerType);
            n.WriteRect(aRect);
            return n;
        }

        public static implicit operator JSONNode(RectOffset aRect)
        {
            JSONNode n = GetContainer(RectContainerType);
            n.WriteRectOffset(aRect);
            return n;
        }

        public static implicit operator JSONNode(Matrix4x4 aMatrix)
        {
            JSONNode n = new JSONArray();
            n.WriteMatrix(aMatrix);
            return n;
        }

        public static implicit operator JSONNode(Color aColor)
        {
            JSONNode n = GetContainer(ColorContainerType);
            n.WriteColor(aColor);
            return n;
        }

        public static implicit operator JSONNode(Color32 aColor32)
        {
            JSONNode n = GetContainer(ColorContainerType);
            n.WriteColor32(aColor32);
            return n;
        }

        public static implicit operator Vector2(JSONNode aNode)
        {
            return aNode.ReadVector2();
        }

        public static implicit operator Vector3(JSONNode aNode)
        {
            return aNode.ReadVector3();
        }

        public static implicit operator Vector4(JSONNode aNode)
        {
            return aNode.ReadVector4();
        }

        public static implicit operator Quaternion(JSONNode aNode)
        {
            return aNode.ReadQuaternion();
        }

        public static implicit operator Rect(JSONNode aNode)
        {
            return aNode.ReadRect();
        }

        public static implicit operator RectOffset(JSONNode aNode)
        {
            return aNode.ReadRectOffset();
        }

        public static implicit operator Matrix4x4(JSONNode aNode)
        {
            return aNode.ReadMatrix();
        }

        public static implicit operator Color(JSONNode aNode)
        {
            return aNode.ReadColor();
        }

        public static implicit operator Color32(JSONNode aNode)
        {
            return aNode.ReadColor32();
        }

        #endregion implicit conversion operators

        #region Vector2
        public Vector2 ReadVector2(Vector2 aDefault)
        {
            if (IsObject)
            {
                return new Vector2(this["x"].AsFloat, this["y"].AsFloat);
            }
            if (IsArray)
            {
                return new Vector2(this[0].AsFloat, this[1].AsFloat);
            }
            return aDefault;
        }

        public Vector2 ReadVector2(string aXName, string aYName)
        {
            if (IsObject)
            {
                return new Vector2(this[aXName].AsFloat, this[aYName].AsFloat);
            }
            return Vector2.zero;
        }

        public Vector2 ReadVector2()
        {
            return ReadVector2(Vector2.zero);
        }

        public JSONNode WriteVector2(Vector2 aVec, string aXName = "x", string aYName = "y")
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this[aXName].AsFloat = aVec.x;
                this[aYName].AsFloat = aVec.y;
            }
            else if (IsArray)
            {
                Inline = true;

                for (int i = 0; i < 2; i++)
                {
                    Add(aVec[i]);
                }
            }
            return this;
        }
        #endregion Vector2

        #region Vector3
        public Vector3 ReadVector3(Vector3 aDefault)
        {
            if (IsObject)
            {
                return new Vector3(this["x"].AsFloat, this["y"].AsFloat, this["z"].AsFloat);
            }
            if (IsArray)
            {
                return new Vector3(this[0].AsFloat, this[1].AsFloat, this[2].AsFloat);
            }
            return aDefault;
        }

        public Vector3 ReadVector3(string aXName, string aYName, string aZName)
        {
            if (IsObject)
            {
                return new Vector3(this[aXName].AsFloat, this[aYName].AsFloat, this[aZName].AsFloat);
            }
            return Vector3.zero;
        }

        public Vector3 ReadVector3()
        {
            return ReadVector3(Vector3.zero);
        }

        public JSONNode WriteVector3(Vector3 aVec, string aXName = "x", string aYName = "y", string aZName = "z")
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this[aXName].AsFloat = aVec.x;
                this[aYName].AsFloat = aVec.y;
                this[aZName].AsFloat = aVec.z;
            }
            else if (IsArray)
            {
                Inline = true;
                for (int i = 0; i < 3; i++)
                {
                    Add(aVec[i]);
                }
            }
            return this;
        }
        #endregion Vector3

        #region Vector4
        public Vector4 ReadVector4(Vector4 aDefault)
        {
            if (IsObject)
            {
                return new Vector4(this["x"].AsFloat, this["y"].AsFloat, this["z"].AsFloat, this["w"].AsFloat);
            }
            if (IsArray)
            {
                return new Vector4(this[0].AsFloat, this[1].AsFloat, this[2].AsFloat, this[3].AsFloat);
            }
            return aDefault;
        }

        public Vector4 ReadVector4()
        {
            return ReadVector4(Vector4.zero);
        }

        public JSONNode WriteVector4(Vector4 aVec)
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this["x"].AsFloat = aVec.x;
                this["y"].AsFloat = aVec.y;
                this["z"].AsFloat = aVec.z;
                this["w"].AsFloat = aVec.w;
            }
            else if (IsArray)
            {
                Inline = true;

                for (int i = 0; i < 4; i++)
                {
                    Add(aVec[i]);
                }
            }
            return this;
        }
        #endregion Vector4

        #region Quaternion
        public Quaternion ReadQuaternion(Quaternion aDefault)
        {
            if (IsObject)
            {
                return new Quaternion(this["x"].AsFloat, this["y"].AsFloat, this["z"].AsFloat, this["w"].AsFloat);
            }
            if (IsArray)
            {
                return new Quaternion(this[0].AsFloat, this[1].AsFloat, this[2].AsFloat, this[3].AsFloat);
            }
            return aDefault;
        }

        public Quaternion ReadQuaternion()
        {
            return ReadQuaternion(Quaternion.identity);
        }

        public JSONNode WriteQuaternion(Quaternion aRot)
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this["x"].AsFloat = aRot.x;
                this["y"].AsFloat = aRot.y;
                this["z"].AsFloat = aRot.z;
                this["w"].AsFloat = aRot.w;
            }
            else if (IsArray)
            {
                Inline = true;

                for (int i = 0; i < 4; i++)
                {
                    Add(aRot[i]);
                }
            }
            return this;
        }
        #endregion Quaternion

        #region Rect
        public Rect ReadRect(Rect aDefault)
        {
            if (IsObject)
            {
                return new Rect(this["x"].AsFloat, this["y"].AsFloat, this["width"].AsFloat, this["height"].AsFloat);
            }
            if (IsArray)
            {
                return new Rect(this[0].AsFloat, this[1].AsFloat, this[2].AsFloat, this[3].AsFloat);
            }
            return aDefault;
        }

        public Rect ReadRect()
        {
            return ReadRect(new Rect());
        }

        public JSONNode WriteRect(Rect aRect)
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this["x"].AsFloat = aRect.x;
                this["y"].AsFloat = aRect.y;
                this["width"].AsFloat = aRect.width;
                this["height"].AsFloat = aRect.height;
            }
            else if (IsArray)
            {
                Inline = true;
                Add(aRect.x);
                Add(aRect.y);
                Add(aRect.width);
                Add(aRect.height);
            }
            return this;
        }
        #endregion Rect

        #region RectOffset
        public RectOffset ReadRectOffset(RectOffset aDefault)
        {
            if (this is JSONObject)
            {
                return new RectOffset(this["left"].AsInt, this["right"].AsInt, this["top"].AsInt, this["bottom"].AsInt);
            }
            if (this is JSONArray)
            {
                return new RectOffset(this[0].AsInt, this[1].AsInt, this[2].AsInt, this[3].AsInt);
            }
            return aDefault;
        }

        public RectOffset ReadRectOffset()
        {
            return ReadRectOffset(new RectOffset());
        }

        public JSONNode WriteRectOffset(RectOffset aRect)
        {
            Clear();

            if (IsObject)
            {
                Inline = true;
                this["left"].AsInt = aRect.left;
                this["right"].AsInt = aRect.right;
                this["top"].AsInt = aRect.top;
                this["bottom"].AsInt = aRect.bottom;
            }
            else if (IsArray)
            {
                Inline = true;
                Add(aRect.left);
                Add(aRect.right);
                Add(aRect.top);
                Add(aRect.bottom);
            }
            return this;
        }
        #endregion RectOffset

        #region Matrix4x4
        public Matrix4x4 ReadMatrix()
        {
            Matrix4x4 result = Matrix4x4.identity;
            if (IsArray)
            {
                for (int i = 0; i < 16; i++)
                {
                    result[i] = this[i].AsFloat;
                }
            }
            return result;
        }

        public JSONNode WriteMatrix(Matrix4x4 aMatrix)
        {
            Clear();

            if (IsArray)
            {
                Inline = true;
                for (int i = 0; i < 16; i++)
                {
                    Add(aMatrix[i]);
                }
            }
            return this;
        }
        #endregion Matrix4x4

        #region Color
        public Color ReadColor()
        {
            if (IsString && ColorUtility.TryParseHtmlString(Value, out Color htmlColor))
            {
                return htmlColor;
            }

            if (IsArray)
            {
                return ReadVector4();
            }

            if (IsObject)
            {
                return new Color(this["r"].AsFloat, this["g"].AsFloat, this["b"].AsFloat, this["a"].AsFloat);
            }

            return Color.white;
        }

        public JSONNode WriteColor(Color aColor)
        {
            Clear();

            if (IsString)
            {
                Value = $"#{ColorUtility.ToHtmlStringRGBA(aColor)}";
            }
            else if (IsObject)
            {
                Inline = true;
                this["r"].AsFloat = aColor.r;
                this["g"].AsFloat = aColor.g;
                this["b"].AsFloat = aColor.b;
                this["a"].AsFloat = aColor.a;
            }
            else if (IsArray)
            {
                WriteVector4(aColor);
            }

            return this;
        }
        #endregion Color

        #region Color32
        public Color32 ReadColor32()
        {
            if (IsString && ColorUtility.TryParseHtmlString(Value, out Color htmlColor))
            {
                return htmlColor;
            }

            if (IsArray)
            {
                return new Color32(this[0].AsByte, this[1].AsByte, this[2].AsByte, this[3].AsByte);
            }

            if (IsObject)
            {
                return new Color32(this["r"].AsByte, this["g"].AsByte, this["b"].AsByte, this["a"].AsByte);
            }

            return Color.white;
        }

        public JSONNode WriteColor32(Color32 aColor32)
        {
            Clear();

            if (IsString)
            {
                Value = $"#{ColorUtility.ToHtmlStringRGBA(aColor32)}";
            }
            else if (IsObject)
            {
                Inline = true;
                this["r"].AsByte = aColor32.r;
                this["g"].AsByte = aColor32.g;
                this["b"].AsByte = aColor32.b;
                this["a"].AsByte = aColor32.a;
            }
            else if (IsArray)
            {
                Add(aColor32.r);
                Add(aColor32.g);
                Add(aColor32.b);
                Add(aColor32.a);
            }

            return this;
        }
        #endregion Color32

        #region Pose
        public static implicit operator JSONNode(Pose aPose)
        {
            JSONNode n = new JSONObject();
            n.WritePose(aPose);
            return n;
        }

        public static implicit operator Pose(JSONNode aNode)
        {
            return aNode.ReadPose();
        }

        public Pose ReadPose()
        {
            if (IsObject)
            {
                return new Pose(this["position"].ReadVector3(),
                    this["rotation"].ReadQuaternion());
            }
            if (IsArray)
            {
                return new Pose(this[0].ReadVector3(),
                    this[1].ReadQuaternion());
            }

            return new Pose();
        }

        public JSONNode WritePose(Pose aPose)
        {
            Clear();

            if (IsObject)
            {
                this["position"] = aPose.position;
                this["rotation"] = aPose.rotation;
            }
            else if (IsArray)
            {
                Add(aPose.position);
                Add(aPose.rotation);
            }

            return this;
        }

        #endregion Pose
    }
}
#endif
